pkgs/development/python-modules/rapidocr-onnxruntime/default.nix

```diff
@@ -0,0 +1,95 @@
+{
+  lib,
+  buildPythonPackage,
+  fetchFromGitHub,
+
+  fetchgit,
+
+  pythonRelaxDepsHook,
+  pytestCheckHook,
+
+  setuptools,
+  pyclipper,
+  opencv4,
+  numpy,
+  six,
+  shapely,
+  pyyaml,
+  pillow,
+  onnxruntime,
+}:
+let
+  version = "1.3.22";
+
+  src = fetchFromGitHub {
+    owner = "RapidAI";
+    repo = "RapidOCR";
+    rev = "v${version}";
+    hash = "sha256-8h4DFhnI9imr+bYQZdlrl8UKUdpwnGK+SGxLXSMmcag=";
+  };
+
+  models = fetchgit {
+    url = "https://huggingface.co/SWHL/RapidOCR";
+    rev = "38f67e0558416db6056ddd8613292d23b5ff002c";
+    hash = "sha256-11hIFqUf3Ecee3XceYix7cXkhko5rz4BfztvSly18ys=";
+    fetchLFS = true;
+  };
+in
+buildPythonPackage {
+  pname = "rapidocr-onnxruntime";
+  inherit version src;
+  pyproject = true;
+
+  sourceRoot = "${src.name}/python";
+
+  # The setup.py supplied by upstream tries to determine the current version by
+  # fetching the latest version of the package from PyPI, and then bumping the version number.
+  # This is not allowed in the Nix build environment as we do not have internet access,
+  # hence we patch that out and get the version from the build environment directly.
+  patches = [ ./setup-py-override-version-checking.patch ];
+
+  nativeBuildInputs = [
+    pythonRelaxDepsHook
+    pytestCheckHook
+  ];
+
+  build-system = [ setuptools ];
+
+  dependencies = [
+    pyclipper
+    opencv4
+    numpy
+    six
+    shapely
+    pyyaml
+    pillow
+    onnxruntime
+  ];
+
+  # Don't use vendored OpenCV
+  pythonRemoveDeps = [ "opencv-python" ];
+
+  # FIXME: Why isn't this working???
+  #pythonImportsCheck = [ "rapidocr_onnxruntime" ];
```

Comment:
I guess this will have to be fixed before merging. What's the error log?

Reply:
Same as before: ModuleNotFoundError: No module named 'rapidocr_onnxruntime'.
Though, I think I found a reason for that — somehow the `__init__,py` and `main.py` are just completely missing from the built wheel. Given this project uses a very unconventional layout, I wonder if we're missing something again...

Reply:
This is a blocker.
```shell
$ nix-shell -p 'python311.withPackages (ps: with ps; [rapidocr-onnxruntime])' -I nixpkgs=. --run 'python -c "import rapidocr_onnxruntime"'

Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'rapidocr_onnxruntime'
```

Reply:
Well, of course it's a blocker, as it doesn't even work :sweat_smile:

Reply:
Regarding `__init__.py`, we need to add the following command to `postPatch`.
```
echo "from .rapidocr_onnxruntime.main import RapidOCR, VisRes" > rapidocr_onnxruntime/__init__.py
```
https://github.com/RapidAI/RapidOCR/blob/02829ef986bc2a5c4f33e9c45c9267bcf2d07a1d/.github/workflows/gen_whl_to_pypi_rapidocr_ort.yml#L86-L87

Reply:
That doesn't quite work with the current setup — tests fail:
```
       > ______________________ ERROR collecting tests/test_ort.py ______________________
       > ImportError while importing test module '/build/source/python/tests/test_ort.py'.
       > Hint: make sure your test modules/packages have valid Python names.
       > Traceback:
       > /nix/store/7hnr99nxrd2aw6lghybqdmkckq60j6l9-python3-3.11.9/lib/python3.11/importlib/__init__.py:126: in import_module
       >     return _bootstrap._gcd_import(name[level:], package, level)
       > tests/test_ort.py:16: in <module>
       >     from rapidocr_onnxruntime import LoadImageError, RapidOCR
       > rapidocr_onnxruntime/__init__.py:1: in <module>
       >     from .rapidocr_onnxruntime.main import RapidOCR, VisRes
       > E   ModuleNotFoundError: No module named 'rapidocr_onnxruntime.rapidocr_onnxruntime'
```

Or with `from .rapidocr_onnxruntime.main import RapidOCR, VisRes`:
```
       > ______________________ ERROR collecting tests/test_ort.py ______________________
       > ImportError while importing test module '/build/source/python/tests/test_ort.py'.
       > Hint: make sure your test modules/packages have valid Python names.
       > Traceback:
       > /nix/store/7hnr99nxrd2aw6lghybqdmkckq60j6l9-python3-3.11.9/lib/python3.11/importlib/__init__.py:126: in import_module
       >     return _bootstrap._gcd_import(name[level:], package, level)
       > tests/test_ort.py:16: in <module>
       >     from rapidocr_onnxruntime import LoadImageError, RapidOCR
       > E   ImportError: cannot import name 'LoadImageError' from 'rapidocr_onnxruntime' (/build/source/python/rapidocr_onnxruntime/__init__.py)
```

Reply:
I suspect the `setuptools` namespace is causing the problem.
https://github.com/RapidAI/RapidOCR/blob/02829ef986bc2a5c4f33e9c45c9267bcf2d07a1d/python/setup_onnxruntime.py#L59-L60

https://setuptools.pypa.io/en/latest/userguide/package_discovery.html

Reply:
> Regarding `__init__.py`, we need to add the following command to `postPatch`.
> 
> ```
> echo "from .rapidocr_onnxruntime.main import RapidOCR, VisRes" > rapidocr_onnxruntime/__init__.py
> ```
> 
> https://github.com/RapidAI/RapidOCR/blob/02829ef986bc2a5c4f33e9c45c9267bcf2d07a1d/.github/workflows/gen_whl_to_pypi_rapidocr_ort.yml#L86-L87

Why god why are they doing that ?!

Reply:
> > Regarding `__init__.py`, we need to add the following command to `postPatch`.
> > ```
> > echo "from .rapidocr_onnxruntime.main import RapidOCR, VisRes" > rapidocr_onnxruntime/__init__.py
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > [RapidAI/RapidOCR@`02829ef`/.github/workflows/gen_whl_to_pypi_rapidocr_ort.yml#L86-L87](https://github.com/RapidAI/RapidOCR/blob/02829ef986bc2a5c4f33e9c45c9267bcf2d07a1d/.github/workflows/gen_whl_to_pypi_rapidocr_ort.yml#L86-L87)
> 
> Why god why are they doing that ?!

Beats me... This package is just built in an unorthodox way in every sense of the word.